framework/audio/devtools/audioenginedevtools.cpp

```diff
@@ -26,69 +30,217 @@ using namespace mu::midi;
+        }
+    }
+}
+
+void AudioEngineDevTools::setLevelNoise(float level)
+{
+    if (m_noiseChannel) {
+        auto channel = audioEngine()->mixer()->channel(m_noiseChannel.value_or(-1));
+        if (channel) {
+            channel->setLevel(level);
+        }
+    }
+}
+
+void AudioEngineDevTools::setBalanceSine(float balance)
+{
+    if (m_sineChannelId) {
+        auto channel = audioEngine()->mixer()->channel(m_sineChannelId.value_or(-1));
+        if (channel) {
+            channel->setBalance(balance);
+        }
+    }
+}
+
+void AudioEngineDevTools::setBalanceNoise(float balance)
+{
+    if (m_noiseChannel) {
+        auto channel = audioEngine()->mixer()->channel(m_noiseChannel.value_or(-1));
+        if (channel) {
+            channel->setBalance(balance);
+        }
+    }
+}
+
+void AudioEngineDevTools::enableNoiseEq(bool enable)
+{
+    if (m_noiseChannel) {
+        auto channel = audioEngine()->mixer()->channel(m_noiseChannel.value_or(-1));
+        if (channel) {
+            channel->insert(0)->setActive(enable);
+        }
+    }
 }
 
-void AudioEngineDevTools::playSourceMidi()
+void AudioEngineDevTools::playNoise()
 {
-    if (!m_midiSource) {
-        m_midiSource = std::make_shared<MidiSource>();
+    auto source = std::make_shared<NoiseSource>();
+    source->setType(NoiseSource::Type::PINK);
+
+    auto eq = std::make_shared<EqInsert>();
+    eq->setActive(false);
+    if (!m_noiseChannel) {
+        m_noiseChannel = audioEngine()->mixer()->addChannel(source);
+        auto channel = audioEngine()->mixer()->channel(m_noiseChannel.value_or(-1));
+        channel->setInsert(0, eq);
     }
+}
 
+void AudioEngineDevTools::stopNoise()
+{
+    audioEngine()->mixer()->removeChannel(m_noiseChannel.value_or(-1));
+    m_noiseChannel.reset();
+}
+
+void AudioEngineDevTools::playSequencerMidi()
+{
     if (!m_midiStream) {
         makeArpeggio();
-        m_midiSource->loadMIDI(m_midiStream);
     }
 
-    m_midiHandel = audioEngine()->play(m_midiSource);
+    sequencer()->setMIDITrack(0, m_midiStream);
+    sequencer()->play();
 }
 
-void AudioEngineDevTools::stopSourceMidi()
+void AudioEngineDevTools::stopSequencerMidi()
 {
-    audioEngine()->stop(m_midiHandel);
+    sequencer()->stop();
 }
 
 void AudioEngineDevTools::playPlayerMidi()
 {
     makeArpeggio();
-
-    player()->setMidiStream(m_midiStream);
-    player()->play();
+    m_threadMIDIPlayer = sequencer()->playMidi(m_midiStream->initData);
 }
 
 void AudioEngineDevTools::stopPlayerMidi()
 {
-    player()->stop();
+    if (auto player = m_threadMIDIPlayer.lock()) {
+        player->stop();
+    }
 }
 
-void AudioEngineDevTools::playNotation()
+void AudioEngineDevTools::play()
 {
     auto notation = globalContext()->currentNotation();
-    if (!notation) {
-        LOGE() << "no notation";
-        return;
+    if (notation) {
+        auto stream = notation->playback()->midiStream();
+        sequencer()->setMIDITrack(0, stream);
+    }
+    sequencer()->rewind();
```

Comment:
If there is no notation, will this code work correctly?

Reply:
@RomanPudashkin 
yes. It's possible that sequencer has audioplayer but not midiplayer